pkg/server/api/entry/v1/service.go

```diff
@@ -169,6 +175,83 @@ func (s *Service) BatchDeleteEntry(ctx context.Context, req *entry.BatchDeleteEn
 	}, nil
 }
 
+func (s *Service) updateEntry(ctx context.Context, e *types.Entry, inputMask *types.EntryMask, outputMask *types.EntryMask) *entry.BatchUpdateEntryResponse_Result {
+	log := rpccontext.Logger(ctx)
+
+	if e == nil {
+		log.Error("missing entry")
+		return &entry.BatchUpdateEntryResponse_Result{
+			Status: api.CreateStatus(codes.Internal, "missing entry"),
+		}
+	}
+
+	if inputMask == nil {
+		log.Error("missing input mask")
+		return &entry.BatchUpdateEntryResponse_Result{
+			Status: api.CreateStatus(codes.Internal, "missing input mask"),
+		}
+	}
+	if e.SpiffeId != nil {
+		log = log.WithField(telemetry.SPIFFEID, e.SpiffeId)
+	}
+	// Fetch the previous version of the entry.
+	res, err := s.ds.FetchRegistrationEntry(ctx,
+		&datastore.FetchRegistrationEntryRequest{EntryId: e.Id},
+	)
+	if err != nil {
+		log.WithError(err).Error("failed to fetch entry")
+		return &entry.BatchUpdateEntryResponse_Result{
+			Status: api.CreateStatus(codes.Internal, "failed to fetch entry"),
+		}
+	}
+	oldEntry := res.Entry
+
+	// In general it's illegal for an entry to have an empty ParentId or SpiffeId.
+	// But, when updating, it's OK for these fields to be empty as long as they aren't being changed.
+	// To prevent an error we put dummy data in the update request (the actual values don't matter).
+	if !inputMask.SpiffeId {
+		e.SpiffeId = &types.SPIFFEID{TrustDomain: "spiffe://dummy_td", Path: "dummy_path"}
+	}
+	if !inputMask.ParentId {
+		e.ParentId = &types.SPIFFEID{TrustDomain: "spiffe://dummy_td", Path: "dummy_path"}
+	}
+	convEntry, err := api.ProtoToRegistrationEntry(e)
+	if err != nil || convEntry == nil {
+		log.WithError(err).Error("failed to convert entry")
+		return &entry.BatchUpdateEntryResponse_Result{
+			Status: api.CreateStatus(codes.InvalidArgument, "failed to convert entry: %v", err),
+		}
+	}
+
+	// Merge the fields in the input mask from the new entry to the old entry
+	newEntry := mergeEntries(oldEntry, convEntry, inputMask)
+
+	resp, err := s.ds.UpdateRegistrationEntry(ctx, &datastore.UpdateRegistrationEntryRequest{
+		Entry: newEntry,
+	})
+	if err != nil {
+		log.WithError(err).Error("failed to update entry")
+		return &entry.BatchUpdateEntryResponse_Result{
+			Status: api.CreateStatus(codes.Internal, "failed to update entry"),
+		}
+	}
+
+	tEntry, err := api.RegistrationEntryToProto(resp.Entry)
+	if err != nil {
+		log.WithError(err).Error("Unable to convert registration entry in updateEntry")
+		return &entry.BatchUpdateEntryResponse_Result{
+			Status: api.CreateStatus(codes.Internal, "Unable to convert registration entry in updateEntry"),
+		}
+	}
+
+	applyMask(tEntry, outputMask)
+
```

Comment:
❓ 
When creating entries, we check a constraint that the new entry is "unique"
("Unique" means unique {parent ID, spiffe ID, selectors})
I didn't do this here, so it's possible to create a unique entry, then update it to be non-unique.
I can easily add a call to isEntryUnique to check the constraint here, but it does add another query... what do you think?

Reply:
Resolving for now... will wait for feedback on the issue in GH